bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/connection/Clip2Bridge.java

```diff
@@ -0,0 +1,1049 @@
+import org.eclipse.jetty.http.HttpURI;
+import org.eclipse.jetty.http.HttpVersion;
+import org.eclipse.jetty.http.MetaData;
+import org.eclipse.jetty.http.MetaData.Response;
+import org.eclipse.jetty.http2.api.Session;
+import org.eclipse.jetty.http2.api.Stream;
+import org.eclipse.jetty.http2.client.HTTP2Client;
+import org.eclipse.jetty.http2.frames.DataFrame;
+import org.eclipse.jetty.http2.frames.GoAwayFrame;
+import org.eclipse.jetty.http2.frames.HeadersFrame;
+import org.eclipse.jetty.http2.frames.PingFrame;
+import org.eclipse.jetty.http2.frames.ResetFrame;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.Promise.Completable;
+import org.eclipse.jetty.util.ssl.SslContextFactory;
+import org.openhab.binding.hue.internal.dto.CreateUserRequest;
+import org.openhab.binding.hue.internal.dto.SuccessResponse;
+import org.openhab.binding.hue.internal.dto.clip2.BridgeConfig;
+import org.openhab.binding.hue.internal.dto.clip2.Event;
+import org.openhab.binding.hue.internal.dto.clip2.Resource;
+import org.openhab.binding.hue.internal.dto.clip2.ResourceReference;
+import org.openhab.binding.hue.internal.dto.clip2.Resources;
+import org.openhab.binding.hue.internal.dto.clip2.enums.ResourceType;
+import org.openhab.binding.hue.internal.exceptions.ApiException;
+import org.openhab.binding.hue.internal.exceptions.HttpUnauthorizedException;
+import org.openhab.binding.hue.internal.handler.Clip2BridgeHandler;
+import org.openhab.core.io.net.http.HttpClientFactory;
+import org.openhab.core.io.net.http.HttpUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonParser;
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * This class handles HTTP and SSE connections to/from a Hue Bridge running CLIP 2.
+ *
+ * It uses the following connection mechanisms:
+ *
+ * <li>The primary communication uses HTTP 2 streams over a shared permanent HTTP 2 session.</li>
+ * <li>The 'registerApplicationKey()' method uses HTTP/1.1 over the OH common Jetty client.</li>
+ * <li>The 'isClip2Supported()' static method uses HTTP/1.1 over the OH common Jetty client via 'HttpUtil'.</li>
+ *
+ * @author Andrew Fiddian-Green - Initial Contribution
+ */
+@NonNullByDefault
+public class Clip2Bridge implements Closeable {
+
+    /**
+     * Base (abstract) adapter for listening to HTTP 2 stream events.
+     *
+     * It implements a CompletableFuture by means of which the caller can wait for the response data to come in. And
+     * which, in the case of fatal errors, gets completed exceptionally.
+     *
+     * It handles the following fatal error events by notifying the containing class:
+     *
+     * <li>onHeaders() HTTP unauthorized codes</li>
+     */
+    private abstract class BaseStreamListenerAdapter<T> extends Stream.Listener.Adapter {
```

Comment:
Since this file is rather large (> 1000 lines) consider extracting this class to another file.